docs/design/019-file-api.md

```diff
@@ -0,0 +1,226 @@
+
+In line with [our commitment to optimize large file handling in k6](https://github.com/grafana/k6/issues/2974), we propose introducing a new `fs` module. This module is intended to offer an intuitive and user-friendly API for file interactions within k6 scripts. We'll also provide some ideas for efficient file handling to minimize memory consumption during k6 execution.
+
+### Context
+Currently, files cannot be opened from within a function executed by a VU, only in the init context.
+
+This is due to k6's design for distributed execution, particularly in the cloud. K6 runs the init context once, gathers resources, including files, and sends them to other instances where VU code runs.
+
+### Assumptions
+- The solution should be native to k6, with the code being in the k6 repository and maintained by the k6 OSS team.
+- The new functionality's API should be intuitive for users familiar with filesystem APIs from other languages and technologies.
+- The module should introduce only essential APIs: read-only operations.
+- The proposed API should not affect file handling within HTTP requests.
+
+### Requirements
+- The solution should allow file interaction during k6 script execution.
+- The solution should resort to asynchronous operations as much as possible.
+- The module should operate seamlessly in local and cloud setups.
+- The module should strive to optimize memory usage when handling files.
+
+### Scope
+
+#### In scope
+- The proposed solution will offer an alternative to the existing `open` function.
+- The solution will provide a File concept that supports read-only operations: reading file chunks into a buffer, reading the entire file content, and file seeking.
+- The proposal could suggest changes to improve overall memory usage in k6 scripts.
+
+#### Out of scope
+- This proposal doesn't aim to solve memory consumption issues when using file content in HTTP requests, although minor improvements might be suggested.
+- Changes to how k6 gathers resources, particularly files, are out of scope (ideas are proposed in the "Future Improvements" section).
+- Despite the proposed API's potential to support them, write operations are beyond this proposal's scope.
+
+## Proposed solution
+
+We suggest implementing a minimalist, experimental file system (`fs`) module based on [Deno's fs module](https://deno.land/api@v1.32.3?s=Deno.open). The new module will allow users to interact with files, separating text and binary files. The module will provide an `open` function that returns a file handle/view for performing read operations.
+
+The initial API will mostly be asynchronous, except for the `open` functionality which will be synchronous due to the current lack of support for `await` operations within the init context.
```

Comment:
I would prefer keep the `openSync` for the first iteration of the module then; so that we ensure a good and intuitive user experience to newcomers.  I would also include an `open` function already which allows the above in prevision for the future where await works in the init context without tricks 😄 

We would drop it once ESM modules are supported, or before the experimental module goes in core, or even once #3020 is ✅.

Maybe there's a middle-ground that avoids going into weird-ish code like you mentioned though 💭

Reply:
>Maybe there's a middle-ground that avoids going into weird-ish code like you mentioned though

I see the risk where we mostly focus on the sync concept and miss making it work with the async. It seems the final solution should not have this sync API so I wouldn't try to implement it.

The proposal workaround is weird, but it has the benefit of not breaking the API when we are ready to move to the final API so it gives us more immediate feedback about issues.

I'm okay with doing the sync if we mostly do a wrapper around the async feature.

Reply:
We have an agreement then 🤝 

If it's possible to do it as a wrapper around the async function indeed, and to have `openSync` effectively be syntactic sugar above the workaround discussed earlier.